### PR TITLE
fix: impl Trait desugaring in trait objects' assoc constraints

### PR DESCRIPTION
**Pull request by mehdiakiki**
Issue: #96529

Right now, the tasks that are done: 
- [x] saving the desugaring into an ìmpl_trait_context`.
- [x] add `impl_trait_id` to `rust_ast::AssocConstraint`.

**In progress:** 
The next tasks are a break out of what I am trying to accomplish. A bit different from the original tasks in order to be as precise as possible and help with mentoring. 
- [ ] create `visit_assoc_constraint` in `def_collector` and create the local def id there by matching with `UniversalInDyn` context.
- [ ] replace the `desugar_to_impl_trait` with the associated constraint check.
- [ ] remove the enum `rust_ast_lowering::ImplTraitContext`and add the vector to `LoweringContext`.
- [ ] use that vector in the relevant places and retrieve it through the use of `with_hir_id_owner`.

**Not started yet.**
- [ ] create a `ResolverAstLowering::get_impl_trait_context(LocalDefId)` to access the `ImplTraitContext` value.
- [ ] fix method parameter documentation for `lower_fn_decl`.

### CODE REVIEW

File: compiler/rustc_resolve/src/def_collector.rs [hunk 1]
```diff
@@ -285,19 +285,36 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
...
                 let parent_def = match self.impl_trait_context {
-                    ImplTraitContext::Universal(item_def) => self.resolver.create_def(
-                        item_def,
-                        node_id,
-                        DefPathData::ImplTrait,
-                        self.expansion.to_expn_id(),
-                        ty.span,
-                    ),
+                    ImplTraitContext::Universal(item_def) => {
+                        //| ImplTraitContext::UniversalInDyn(item_def) => {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/def_collector.rs:289]:
```suggestion
                        | ImplTraitContext::UniversalInDyn(item_def) => {
```
This is the correct behaviour here.

File: compiler/rustc_resolve/src/def_collector.rs [hunk 2]
```diff
@@ -285,19 +285,36 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
...
+                        let def_id = self.resolver.create_def(
+                            item_def,
+                            node_id,
+                            DefPathData::ImplTrait,
+                            self.expansion.to_expn_id(),
+                            ty.span,
+                        );
+                        self.resolver
+                            .impl_trait_context
+                            .insert(def_id, ImplTraitContext::Universal(item_def));
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/def_collector.rs:299]:
```suggestion
                            .insert(def_id, self.impl_trait_context);
```
We have not exited the trait object.

File: compiler/rustc_resolve/src/def_collector.rs [hunk 3]
```diff
@@ -285,19 +285,36 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
...
+                            .insert(def_id, ImplTraitContext::Existential);
+                        def_id
                     }
+                    ImplTraitContext::UniversalInDyn(_) => self.parent_def,
                 };
                 self.with_parent(parent_def, |this| visit::walk_ty(this, ty))
             }
+            TyKind::TraitObject(ref bounds, ..) => {
+                self.with_impl_trait(ImplTraitContext::UniversalInDyn(self.parent_def), |this| {
+                    walk_list!(this, visit_param_bound, bounds, BoundKind::TraitObject);
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/def_collector.rs:315]:
```suggestion
                    visit::walk_ty(self, ty)
```
No need to re-code the default behaviour.

File: compiler/rustc_resolve/src/def_collector.rs [hunk 4]
```diff
@@ -352,4 +369,29 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
             visit::walk_crate(self, krate)
         }
     }
+
+    fn visit_assoc_constraint(&mut self, constraint: &'a AssocConstraint) {
+        /*if let ImplTraitContext::UniversalInDyn(item_def) = self.impl_trait_context {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/def_collector.rs:374]: This version is correct.

Comment (cjgillot) [reply compiler/rustc_resolve/src/def_collector.rs:374]: We should only create the definition is `constraint.kind` is `AssocConstraintKind::Bound`.
`dyn Foo<T: Trait>` becomes `dyn Foo<T = impl Trait>`.

File: compiler/rustc_resolve/src/def_collector.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_resolve/src/def_collector.rs:299]: My bad, your code is right.

**Comment by Dylan-DPC** [comment]
@randomicon00 what's the status of this?

**Comment by randomicon00** [comment]
@Dylan-DPC Will have an update in the next two weeks.

**Comment by Dylan-DPC** [comment]
@randomicon00 any updates on this?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**